### DISCUSSION TITLE
Post-update Comprehensive Catalogue takes too long

### DISCUSSION THREAD

**Discussion by vix35** [aws-solutions-library-samples/data-lakes-on-aws#210] [Q&A]
Hi SDLF community, and sorry if there is many mistakes in my english.

I am implementing SDLF with my team, and we are at the point of running the complete process from raw to post-stage, the problem i encountered as i mention on the title of the discussion is that the post update takes too long, and the lambda timeouts before completing all the logging.

here is some images that could help with some more context:

image of the failed step function and parameters:
![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/674a4435-3186-479e-a2d4-a4d3c62f04a8)

Image of the size of the bucket where is logging the objects:
![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/ed15c965-5402-4098-8b90-5fd93affcdaa)

Image of the time it takes between each iterarion of the logging and the code that prints, as you can see it takes near to 300ms for each file to be logged in Dynamo, and inside the "folder" of the bucket which is logging there is more than 3600 objects, and there is not even half of the whole data we want to put, so i was wondering what we are doing wrong here.
![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/fa562386-308a-4748-80ef-0d4d89315dc0)
![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/acc0027a-c81a-496a-9229-c7b240d3c257)

**Comment by cnfait** [comment]
Hello @vix35 

Instead of updating the DynamoDB table record by record, we can probably batch them:
https://github.com/awslabs/aws-serverless-data-lake-framework/commit/9a178cdb6a3b48329e2098b73f3c9a93064bebca

Can you try and let me know if that helps? I'll run a few tests on my side as well. This may be added to SDLF by default.

**Reply by vix35** [reply]
Thanks @cnfait this will absolutely take out some time off the process, when i try it i will tell you how it goes.

I am still a bit worried about the time it takes the process involving the S3Interface, the `"size"` and the `"last_modified_date"` , as you can see in the logs both of that lines of code takes most of the time in each cycle of the `for`.

Thanks for taking the time and effort to help.

**Reply by cnfait** [reply]
Indeed. These two calls are each sending a HEAD request to S3 - fetching the same metadata, twice. There might be an opportunity for optimization there but I need to verify a few assumptions. I'll get back to you tomorrow (CET).

**Reply by cnfait** [reply]
Here is another change you can try:
https://github.com/awslabs/aws-serverless-data-lake-framework/commit/b731b4bc3999d446361f105c6a2909c755aa9d65

Like we discussed - I'm fetching metadata from S3 only once. In my tests, total time is at least halved. One more thing you could do if that's not the case already is to configure this lambda to use Python 3.11. There should be no compatibility issue.

Can you try all of these things and report on the improvements? let me know if there are issues, too.

**Reply by vix35** [reply]
Hi @cnfait i took 3 tests:

the post-update as it came originally: 625.679 [ms]
With only accessing once to the metadata in S3: 317.860,48 [ms]
With only accessing once to the metadata in S3 and Dynamo Batch: 297.777

![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/e613d84e-f232-45cf-9c73-1eecf3e7b629)

**Reply by cnfait** [reply]
That's a noticeable improvement, thank you for your tests. Can I also ask which version of Python is used for this lambda in your environment?

**Reply by vix35** [reply]
Python 3.9

**Reply by cnfait** [reply]
[Python 3.11 should be faster](https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-faster-cpython), if you have the opportunity to try it. I don't expect a major speedup as most of the duration is due to network calls but if your use case is still failing this might help as well.

I'm also thinking about other ways to handle pipeline metadata updates (an asynchronous process executed entirely outside pipelines) but this is not even at the draft stage at this point.

Also stageb is usually limited to processing 100 items at a time by default. This is admittedly very low and configurable but I'm interested to know if these thousands of objects **have** to be processed by the glue job at the same time of if they could be processed in multiple glue job runs?

**Reply by vix35** [reply]
The stem of the problem probably was partly our fault, we were merging data in delta (delta.io) tables without running OPTIMIZE, and this process makes A LOT of little files (1 MB), we are planning to run optimize now, the way still is not clear yet, but we are thinking of a daily process (glue) which OPTIMIZE every table once per day as recommended by [databricks](https://learn.microsoft.com/en-us/azure/databricks/delta/optimize#how-often-should-i-run-optimize)

Still i feel the optimization you helped us to get will be helpful in the future when we have more files even with OPTIMIZE running on a daily basis, but if the data goes up the amount of files will always increase, so its bound to fail sometime in the future.

Another thing we did was modify the heavy-transform layer and the post-update lambda to instead of getting the metadata of the whole post-stage, only get the metadata of the tables which were actually changed in the execution of the stage B 

![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/e3a6b911-bb3c-4553-aca2-5678991efdc4)
![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/f9522046-0586-4ad6-a228-762fa94f916a)


As you can see in this image the "tables" list was wrong, and was picking the name of the files with the data of the tables, so we changed it to take the actual name of the tables that were processed and use it to only update the metadata of the files of those tables.

![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/930ffc7b-af90-44a9-909d-dab196929bbb)

Result with the modifications

![image](https://github.com/awslabs/aws-serverless-data-lake-framework/assets/26469760/3957e26f-077c-461f-984e-503bb81484a7)
